File: 4-kyu/snail/snail.js
snail = function(array) {
  const result = [];
  while (array.length) {
    // take top row
    result.push(...array.shift());
    // take right column
    array.forEach(row => result.push(row.pop()));
    // take bottom row (reversed)
    if (array.length) result.push(...array.pop().reverse());
    // take left column (bottom to top)
    for (let i = array.length - 1; i >= 0; i--) {
      if (array[i].length) result.push(array[i].shift());
    }
  }
  return result;
}

File: 4-kyu/snail/snail.py
def snail(snail_map):
    result = []
    while snail_map:
        # take the top row
        result += snail_map.pop(0)
        # take the right column
        if snail_map and snail_map[0]:
            for row in snail_map:
                result.append(row.pop())
        # take the bottom row (reversed)
        if snail_map:
            result += snail_map.pop()[::-1]
        # take the left column (reversed)
        if snail_map and snail_map[0]:
            for row in snail_map[::-1]:
                result.append(row.pop(0))
    return result